wallet/src/Cardano/Wallet/Kernel/DB/HdWallet.hs

```diff
@@ -535,6 +538,16 @@ instance IsOurs [(HdRootId, Core.EncryptedSecretKey)] where
         let addrId = HdAddressId accId addressIx
         return $ HdAddress addrId (InDb addr)
 
+instance IsOurs [ (HdRootId, WalletDecrCredentials) ] where
+    isOurs addr s = (,s) $ foldl' (<|>) Nothing $ flip map s $ \(rootId, wdc) -> do
+        case decryptAddress wdc (addr::Core.Address) of
+          Just (WAddressMeta _wid accountIx addressIx _addrv1) -> do
+            let
+              accId = HdAccountId rootId (HdAccountIx accountIx)
+              addrId = HdAddressId accId (HdAddressIx addressIx)
+            pure (HdAddress addrId (InDb addr))
+          Nothing -> Nothing
```

Comment:
Okay, I see where the gain in time comes from here. `decryptHdLvl2DerivationPath` does pretty much the same thing `decryptAddress` does except that, it has one extra step of computing the `HdPassphrase` from the `HdRootId` (which is a PBKDF2 with 500 iterations, so it takes some time..).